# Black Scholes Vanilla European Call Prices

## Imports

In [1]:
import datetime
y=datetime.datetime.now()
print(y.strftime("%H:%M:%S:%f %A %d %B %Y %z" ))

14:21:07:326164 Sunday 23 December 2018 


In [2]:
import os
print("Current working directory",os.getcwd())

Current working directory C:\Users\Peter\Documents\code\lbs\learningBlackScholes


In [3]:
import numpy as np
np.__version__

'1.15.4'

In [4]:
import pandas as pd
pd.__version__

'0.23.4'

In [5]:
import matplotlib
import matplotlib.pyplot as plt

In [6]:
from math import log, sqrt, exp
from scipy import stats

In [7]:
import decimal
from decimal import Decimal
decimal.getcontext()

Context(prec=28, rounding=ROUND_HALF_EVEN, Emin=-999999, Emax=999999, capitals=1, clamp=0, flags=[], traps=[InvalidOperation, DivisionByZero, Overflow])

In [8]:
decimal.getcontext().prec=112
#increase the number precision

## Functions

In [9]:
'''
Black-Scholes vanilla European call price function, with the default values as shown in the example in Hull (2009)
'''

def bs(S=42,K=40,r=0.1,sigma=0.2,T=0.5):
    if ( np.any(S)==0 ):
        return 0
    elif (np.any(K)==0):
        return S    
    else:    
        d1 = (np.log(S/K)  +   (r+(np.square(sigma)/2) )* T )  / (sigma*np.sqrt(T))  
        d2 = d1 - (sigma * np.sqrt(T) )   
        c = S*stats.norm.cdf(d1, 0.0, 1.0)-  K*np.exp(-r*T)*stats.norm.cdf(d2, 0.0, 1.0)
        return c  

In [10]:
'''
Black-Scholes vanilla European call price function, with the default values as shown in the example in Hull (2009)
'''

def bsDecimal(S=42,K=40,r=0.1,sigma=0.2,T=0.5):
    
    S=Decimal(S)
    K=Decimal(K)
    r=Decimal(r)
    sigma=Decimal(sigma)
    T=Decimal(T)
    
    if ( np.any(S)==0 ):
        return 0
    elif (np.any(K)==0):
        return S    
    else:    
        d1 = (  ( Decimal (S/K) ).ln()  +   Decimal ( (r+(np.square(sigma)/2) )* T ) )  / Decimal ( (sigma*np.sqrt(T)) ) 
        d2 = d1 - Decimal ( (sigma * np.sqrt(T) )   )
        
        '''
        print(d1)
        print(round(d1,20))
        print(float(round(d1,20)))
        print()
        print(d2)
        print(round(d2,20))
        print(float(round(d2,20)))
        '''
        
        #c = Decimal( S* Decimal( stats.norm.cdf(d1, 0.0, 1.0) )  )\
        #            - Decimal (K * (  np.exp(-r*T)*Decimal(stats.norm.cdf(d2, 0.0, 1.0))) )
        
        round_d1=float(round(d1,20))
        print(type(round_d1))
        round_d2=float(round(d2,20))
        print(type(round_d1))
        
        cdf_d1=Decimal ( stats.norm.cdf(round_d1, 0.0, 1.0) ) 
        cdf_d2=Decimal ( stats.norm.cdf(round_d2, 0.0, 1.0) )
        
        c = Decimal (
        
        (
            S*cdf_d1 
        )
            -
        (
            K*np.exp(-r*T)*cdf_d2
        )
        )
        
        return c  

In [11]:
x=Decimal(42)

In [12]:
bsDecimal(S=x)

<class 'float'>
<class 'float'>


Decimal('4.75942239287153143007131791898744713680478009470034327250796964568769722921726499194780886164081340434109854130')

In [13]:
y=float(10)

In [14]:
x+Decimal(y)

Decimal('52')

In [15]:
type(y)

float

In [16]:
bs(S=42)

4.759422392871532

In [90]:
%prun for i in (0,1, 10000): bs(S=i)

#### Function bs() typically produces the Euro call value of 4.759422392871532

In [17]:
bs(K=0)

42

In [18]:
type(y)

float

Hull's value for the default is 4.759, same as the output of the bs function with the default parameters

In [19]:
'''
BSM Call Function taken from Hilpisch (2014), Chapter 10
'''
def bsm_call_value(S0, K, T, r, sigma):
    S0 = float(S0)
    d1 = (log(S0 / K) + (r + 0.5 * sigma ** 2) * T) / (sigma * sqrt(T))
    d2 = (log(S0 / K) + (r - 0.5 * sigma ** 2) * T) / (sigma * sqrt(T))
    value = (S0 * stats.norm.cdf(d1, 0.0, 1.0)
             - K * exp(-r * T) * stats.norm.cdf(d2, 0.0, 1.0))
    return value

# stats.norm.cdf --> cumulative distribution function
#                    for normal distribution

In [20]:
bsm_call_value(42,40,0.5,0.1,0.2)

4.759422392871532

## Parameters for the model

In [21]:
NUMBER_S=10

In [22]:
INCREMENT_S=( 1/1 ) 

In [23]:
INCREMENT_S

1.0

In [24]:
NUMBER_X=10

In [25]:
INCREMENT_X=1/1

In [26]:
INCREMENT_X

1.0

In [27]:
NUMBER_SIGMA=10

In [28]:
INCREMENT_SIGMA=1/1

In [29]:
INCREMENT_SIGMA

1.0

In [183]:
NUMBER_T=252

In [214]:
INCREMENT_T=-21

In [215]:
INCREMENT_T

-21

In [216]:
252/12

21.0

In [217]:
optionValues=np.zeros(NUMBER_S*NUMBER_X*NUMBER_SIGMA)

In [218]:
optionValues.shape

(1000,)

In [219]:
#for m in np.arange (1, 26, 0.1):
#    print("{:1.1f} \t {:3.6f}".format(m,bs(m)))    

In [220]:
optionS=np.array(np.arange(1,NUMBER_S+1,INCREMENT_S))

In [221]:
optionX=np.array(np.arange(1,NUMBER_X+1,INCREMENT_X))

In [222]:
optionSIGMA=np.array(np.arange(1,NUMBER_SIGMA+1,INCREMENT_SIGMA))

In [233]:
optionT=(np.arange(-INCREMENT_T,NUMBER_T-INCREMENT_T,-INCREMENT_T))

In [234]:
optionT

array([ 21,  42,  63,  84, 105, 126, 147, 168, 189, 210, 231, 252])

In [241]:
optionS

array([ 1.,  2.,  3.,  4.,  5.,  6.,  7.,  8.,  9., 10.])

In [242]:
print("{:2.100}".format(optionS[6]/3))

2.333333333333333481363069950020872056484222412109375


In [243]:
def myFunction(x):
    return (1+ ( 2*x) )

In [244]:
NUMBER_S*NUMBER_X*NUMBER_SIGMA*int(-NUMBER_T/INCREMENT_T)

12000

In [245]:
optionValue=np.zeros(NUMBER_S*NUMBER_X*NUMBER_SIGMA*int(-NUMBER_T/INCREMENT_T))

In [246]:
optionValue.shape

(12000,)

In [250]:
%%time
counter=0
for s in optionS:
    for x in optionX:
        for sig in optionSIGMA:
            for t in optionT:
                optionValue[counter]=bs(S=s,K=x,sigma=sig,T=t/252)
                print("{:3.2f}\t{:3.2f}\t{:3.2f}\t{:3f}\t{:3.5f}\n".format(s,x,sig,t,optionValue[counter]))
                counter+=1    


1.00	1.00	1.00	21.000000	0.11849

1.00	1.00	1.00	42.000000	0.16880

1.00	1.00	1.00	63.000000	0.20756

1.00	1.00	1.00	84.000000	0.24020

1.00	1.00	1.00	105.000000	0.26885

1.00	1.00	1.00	126.000000	0.29462

1.00	1.00	1.00	147.000000	0.31816

1.00	1.00	1.00	168.000000	0.33990

1.00	1.00	1.00	189.000000	0.36016

1.00	1.00	1.00	210.000000	0.37914

1.00	1.00	1.00	231.000000	0.39703

1.00	1.00	1.00	252.000000	0.41396

1.00	1.00	2.00	21.000000	0.23040

1.00	1.00	2.00	42.000000	0.32262

1.00	1.00	2.00	63.000000	0.39065

1.00	1.00	2.00	84.000000	0.44570

1.00	1.00	2.00	105.000000	0.49219

1.00	1.00	2.00	126.000000	0.53246

1.00	1.00	2.00	147.000000	0.56792

1.00	1.00	2.00	168.000000	0.59952

1.00	1.00	2.00	189.000000	0.62792

1.00	1.00	2.00	210.000000	0.65361

1.00	1.00	2.00	231.000000	0.67700

1.00	1.00	2.00	252.000000	0.69836

1.00	1.00	3.00	21.000000	0.33777

1.00	1.00	3.00	42.000000	0.46421

1.00	1.00	3.00	63.000000	0.55240

1.00	1.00	3.00	84.000000	0.61995

1.00	1.00	3.00	105.000000	0.6739


1.00	4.00	6.00	252.000000	0.99496

1.00	4.00	7.00	21.000000	0.44903

1.00	4.00	7.00	42.000000	0.71796

1.00	4.00	7.00	63.000000	0.84997

1.00	4.00	7.00	84.000000	0.91831

1.00	4.00	7.00	105.000000	0.95483

1.00	4.00	7.00	126.000000	0.97475

1.00	4.00	7.00	147.000000	0.98577

1.00	4.00	7.00	168.000000	0.99193

1.00	4.00	7.00	189.000000	0.99540

1.00	4.00	7.00	210.000000	0.99737

1.00	4.00	7.00	231.000000	0.99849

1.00	4.00	7.00	252.000000	0.99913

1.00	4.00	8.00	21.000000	0.55315

1.00	4.00	8.00	42.000000	0.80861

1.00	4.00	8.00	63.000000	0.91393

1.00	4.00	8.00	84.000000	0.96021

1.00	4.00	8.00	105.000000	0.98128

1.00	4.00	8.00	126.000000	0.99109

1.00	4.00	8.00	147.000000	0.99572

1.00	4.00	8.00	168.000000	0.99793

1.00	4.00	8.00	189.000000	0.99899

1.00	4.00	8.00	210.000000	0.99951

1.00	4.00	8.00	231.000000	0.99976

1.00	4.00	8.00	252.000000	0.99988

1.00	4.00	9.00	21.000000	0.64540

1.00	4.00	9.00	42.000000	0.87512

1.00	4.00	9.00	63.000000	0.95340

1.00	4.00	9.00	84.000000	0.982


1.00	7.00	3.00	189.000000	0.57752

1.00	7.00	3.00	210.000000	0.62449

1.00	7.00	3.00	231.000000	0.66607

1.00	7.00	3.00	252.000000	0.70288

1.00	7.00	4.00	21.000000	0.05184

1.00	7.00	4.00	42.000000	0.20019

1.00	7.00	4.00	63.000000	0.34529

1.00	7.00	4.00	84.000000	0.46759

1.00	7.00	4.00	105.000000	0.56733

1.00	7.00	4.00	126.000000	0.64797

1.00	7.00	4.00	147.000000	0.71310

1.00	7.00	4.00	168.000000	0.76576

1.00	7.00	4.00	189.000000	0.80843

1.00	7.00	4.00	210.000000	0.84306

1.00	7.00	4.00	231.000000	0.87124

1.00	7.00	4.00	252.000000	0.89422

1.00	7.00	5.00	21.000000	0.13199

1.00	7.00	5.00	42.000000	0.36030

1.00	7.00	5.00	63.000000	0.53629

1.00	7.00	5.00	84.000000	0.66335

1.00	7.00	5.00	105.000000	0.75456

1.00	7.00	5.00	126.000000	0.82028

1.00	7.00	5.00	147.000000	0.86790

1.00	7.00	5.00	168.000000	0.90257

1.00	7.00	5.00	189.000000	0.92793

1.00	7.00	5.00	210.000000	0.94655

1.00	7.00	5.00	231.000000	0.96027

1.00	7.00	5.00	252.000000	0.97041

1.00	7.00	6.00	21.000000	0.

1.00	9.00	9.00	84.000000	0.97423

1.00	9.00	9.00	105.000000	0.98981

1.00	9.00	9.00	126.000000	0.99592

1.00	9.00	9.00	147.000000	0.99836

1.00	9.00	9.00	168.000000	0.99933

1.00	9.00	9.00	189.000000	0.99973

1.00	9.00	9.00	210.000000	0.99989

1.00	9.00	9.00	231.000000	0.99995

1.00	9.00	9.00	252.000000	0.99998

1.00	9.00	10.00	21.000000	0.62979

1.00	9.00	10.00	42.000000	0.88979

1.00	9.00	10.00	63.000000	0.96590

1.00	9.00	10.00	84.000000	0.98917

1.00	9.00	10.00	105.000000	0.99650

1.00	9.00	10.00	126.000000	0.99886

1.00	9.00	10.00	147.000000	0.99962

1.00	9.00	10.00	168.000000	0.99987

1.00	9.00	10.00	189.000000	0.99996

1.00	9.00	10.00	210.000000	0.99999

1.00	9.00	10.00	231.000000	1.00000

1.00	9.00	10.00	252.000000	1.00000

1.00	10.00	1.00	21.000000	0.00000

1.00	10.00	1.00	42.000000	0.00000

1.00	10.00	1.00	63.000000	0.00000

1.00	10.00	1.00	84.000000	0.00002

1.00	10.00	1.00	105.000000	0.00011

1.00	10.00	1.00	126.000000	0.00040

1.00	10.00	1.00	147.000000	0.00104

1.00	10.00


2.00	2.00	1.00	147.000000	0.63632

2.00	2.00	1.00	168.000000	0.67980

2.00	2.00	1.00	189.000000	0.72031

2.00	2.00	1.00	210.000000	0.75829

2.00	2.00	1.00	231.000000	0.79407

2.00	2.00	1.00	252.000000	0.82792

2.00	2.00	2.00	21.000000	0.46080

2.00	2.00	2.00	42.000000	0.64523

2.00	2.00	2.00	63.000000	0.78130

2.00	2.00	2.00	84.000000	0.89139

2.00	2.00	2.00	105.000000	0.98438

2.00	2.00	2.00	126.000000	1.06492

2.00	2.00	2.00	147.000000	1.13584

2.00	2.00	2.00	168.000000	1.19903

2.00	2.00	2.00	189.000000	1.25583

2.00	2.00	2.00	210.000000	1.30723

2.00	2.00	2.00	231.000000	1.35399

2.00	2.00	2.00	252.000000	1.39673

2.00	2.00	3.00	21.000000	0.67554

2.00	2.00	3.00	42.000000	0.92842

2.00	2.00	3.00	63.000000	1.10481

2.00	2.00	3.00	84.000000	1.23989

2.00	2.00	3.00	105.000000	1.34796

2.00	2.00	3.00	126.000000	1.43666

2.00	2.00	3.00	147.000000	1.51069

2.00	2.00	3.00	168.000000	1.57322

2.00	2.00	3.00	189.000000	1.62651

2.00	2.00	3.00	210.000000	1.67224

2.00	2.00	3.00	231.000000	1


2.00	6.00	4.00	105.000000	1.36574

2.00	6.00	4.00	126.000000	1.49133

2.00	6.00	4.00	147.000000	1.59001

2.00	6.00	4.00	168.000000	1.66820

2.00	6.00	4.00	189.000000	1.73057

2.00	6.00	4.00	210.000000	1.78060

2.00	6.00	4.00	231.000000	1.82091

2.00	6.00	4.00	252.000000	1.85351

2.00	6.00	5.00	21.000000	0.55763

2.00	6.00	5.00	42.000000	1.01933

2.00	6.00	5.00	63.000000	1.31580

2.00	6.00	5.00	84.000000	1.51444

2.00	6.00	5.00	105.000000	1.65135

2.00	6.00	5.00	126.000000	1.74751

2.00	6.00	5.00	147.000000	1.81595

2.00	6.00	5.00	168.000000	1.86515

2.00	6.00	5.00	189.000000	1.90078

2.00	6.00	5.00	210.000000	1.92674

2.00	6.00	5.00	231.000000	1.94575

2.00	6.00	5.00	252.000000	1.95973

2.00	6.00	6.00	21.000000	0.78600

2.00	6.00	6.00	42.000000	1.28427

2.00	6.00	6.00	63.000000	1.56172

2.00	6.00	6.00	84.000000	1.72560

2.00	6.00	6.00	105.000000	1.82569

2.00	6.00	6.00	126.000000	1.88812

2.00	6.00	6.00	147.000000	1.92764

2.00	6.00	6.00	168.000000	1.95291

2.00	6.00	6.00	189.000000	1

2.00	10.00	6.00	147.000000	1.90886

2.00	10.00	6.00	168.000000	1.94053

2.00	10.00	6.00	189.000000	1.96103

2.00	10.00	6.00	210.000000	1.97437

2.00	10.00	6.00	231.000000	1.98309

2.00	10.00	6.00	252.000000	1.98882

2.00	10.00	7.00	21.000000	0.81838

2.00	10.00	7.00	42.000000	1.38537

2.00	10.00	7.00	63.000000	1.67080

2.00	10.00	7.00	84.000000	1.82006

2.00	10.00	7.00	105.000000	1.90025

2.00	10.00	7.00	126.000000	1.94414

2.00	10.00	7.00	147.000000	1.96847

2.00	10.00	7.00	168.000000	1.98210

2.00	10.00	7.00	189.000000	1.98979

2.00	10.00	7.00	210.000000	1.99415

2.00	10.00	7.00	231.000000	1.99664

2.00	10.00	7.00	252.000000	1.99806

2.00	10.00	8.00	21.000000	1.03516

2.00	10.00	8.00	42.000000	1.58098

2.00	10.00	8.00	63.000000	1.81046

2.00	10.00	8.00	84.000000	1.91210

2.00	10.00	8.00	105.000000	1.95856

2.00	10.00	8.00	126.000000	1.98023

2.00	10.00	8.00	147.000000	1.99049

2.00	10.00	8.00	168.000000	1.99539

2.00	10.00	8.00	189.000000	1.99776

2.00	10.00	8.00	210.000000	1.99890




3.00	4.00	3.00	147.000000	2.15475

3.00	4.00	3.00	168.000000	2.26241

3.00	4.00	3.00	189.000000	2.35425

3.00	4.00	3.00	210.000000	2.43313

3.00	4.00	3.00	231.000000	2.50123

3.00	4.00	3.00	252.000000	2.56029

3.00	4.00	4.00	21.000000	1.07563

3.00	4.00	4.00	42.000000	1.58600

3.00	4.00	4.00	63.000000	1.91936

3.00	4.00	4.00	84.000000	2.15731

3.00	4.00	4.00	105.000000	2.33448

3.00	4.00	4.00	126.000000	2.46975

3.00	4.00	4.00	147.000000	2.57474

3.00	4.00	4.00	168.000000	2.65721

3.00	4.00	4.00	189.000000	2.72255

3.00	4.00	4.00	210.000000	2.77468

3.00	4.00	4.00	231.000000	2.81649

3.00	4.00	4.00	252.000000	2.85019

3.00	4.00	5.00	21.000000	1.38974

3.00	4.00	5.00	42.000000	1.94879

3.00	4.00	5.00	63.000000	2.27951

3.00	4.00	5.00	84.000000	2.49394

3.00	4.00	5.00	105.000000	2.63911

3.00	4.00	5.00	126.000000	2.73993

3.00	4.00	5.00	147.000000	2.81113

3.00	4.00	5.00	168.000000	2.86202

3.00	4.00	5.00	189.000000	2.89872

3.00	4.00	5.00	210.000000	2.92537

3.00	4.00	5.00	231.000000	2


3.00	6.00	5.00	168.000000	2.83232

3.00	6.00	5.00	189.000000	2.87681

3.00	6.00	5.00	210.000000	2.90916

3.00	6.00	5.00	231.000000	2.93280

3.00	6.00	5.00	252.000000	2.95016

3.00	6.00	6.00	21.000000	1.42683

3.00	6.00	6.00	42.000000	2.09223

3.00	6.00	6.00	63.000000	2.44922

3.00	6.00	6.00	84.000000	2.65698

3.00	6.00	6.00	105.000000	2.78285

3.00	6.00	6.00	126.000000	2.86097

3.00	6.00	6.00	147.000000	2.91024

3.00	6.00	6.00	168.000000	2.94168

3.00	6.00	6.00	189.000000	2.96191

3.00	6.00	6.00	210.000000	2.97502

3.00	6.00	6.00	231.000000	2.98356

3.00	6.00	6.00	252.000000	2.98915

3.00	6.00	7.00	21.000000	1.71977

3.00	6.00	7.00	42.000000	2.36756

3.00	6.00	7.00	63.000000	2.66860

3.00	6.00	7.00	84.000000	2.82109

3.00	6.00	7.00	105.000000	2.90162

3.00	6.00	7.00	126.000000	2.94522

3.00	6.00	7.00	147.000000	2.96922

3.00	6.00	7.00	168.000000	2.98258

3.00	6.00	7.00	189.000000	2.99009

3.00	6.00	7.00	210.000000	2.99433

3.00	6.00	7.00	231.000000	2.99675

3.00	6.00	7.00	252.000000	2


3.00	10.00	2.00	126.000000	0.76242

3.00	10.00	2.00	147.000000	0.90290

3.00	10.00	2.00	168.000000	1.03388

3.00	10.00	2.00	189.000000	1.15578

3.00	10.00	2.00	210.000000	1.26917

3.00	10.00	2.00	231.000000	1.37466

3.00	10.00	2.00	252.000000	1.47287

3.00	10.00	3.00	21.000000	0.16922

3.00	10.00	3.00	42.000000	0.52250

3.00	10.00	3.00	63.000000	0.84961

3.00	10.00	3.00	84.000000	1.13090

3.00	10.00	3.00	105.000000	1.37068

3.00	10.00	3.00	126.000000	1.57554

3.00	10.00	3.00	147.000000	1.75134

3.00	10.00	3.00	168.000000	1.90288

3.00	10.00	3.00	189.000000	2.03405

3.00	10.00	3.00	210.000000	2.14801

3.00	10.00	3.00	231.000000	2.24732

3.00	10.00	3.00	252.000000	2.33413

3.00	10.00	4.00	21.000000	0.44081

3.00	10.00	4.00	42.000000	1.00356

3.00	10.00	4.00	63.000000	1.43192

3.00	10.00	4.00	84.000000	1.75720

3.00	10.00	4.00	105.000000	2.00772

3.00	10.00	4.00	126.000000	2.20310

3.00	10.00	4.00	147.000000	2.35701

3.00	10.00	4.00	168.000000	2.47919

3.00	10.00	4.00	189.000000	2.57681


4.00	2.00	7.00	189.000000	3.99341

4.00	2.00	7.00	210.000000	3.99623

4.00	2.00	7.00	231.000000	3.99784

4.00	2.00	7.00	252.000000	3.99876

4.00	2.00	8.00	21.000000	3.31905

4.00	2.00	8.00	42.000000	3.71718

4.00	2.00	8.00	63.000000	3.87442

4.00	2.00	8.00	84.000000	3.94237

4.00	2.00	8.00	105.000000	3.97301

4.00	2.00	8.00	126.000000	3.98719

4.00	2.00	8.00	147.000000	3.99386

4.00	2.00	8.00	168.000000	3.99703

4.00	2.00	8.00	189.000000	3.99856

4.00	2.00	8.00	210.000000	3.99930

4.00	2.00	8.00	231.000000	3.99966

4.00	2.00	8.00	252.000000	3.99983

4.00	2.00	9.00	21.000000	3.46578

4.00	2.00	9.00	42.000000	3.81681

4.00	2.00	9.00	63.000000	3.93238

4.00	2.00	9.00	84.000000	3.97413

4.00	2.00	9.00	105.000000	3.98989

4.00	2.00	9.00	126.000000	3.99599

4.00	2.00	9.00	147.000000	3.99839

4.00	2.00	9.00	168.000000	3.99935

4.00	2.00	9.00	189.000000	3.99974

4.00	2.00	9.00	210.000000	3.99989

4.00	2.00	9.00	231.000000	3.99996

4.00	2.00	9.00	252.000000	3.99998

4.00	2.00	10.00	21.000000	3.

4.00	4.00	8.00	189.000000	3.99795

4.00	4.00	8.00	210.000000	3.99900

4.00	4.00	8.00	231.000000	3.99951

4.00	4.00	8.00	252.000000	3.99976

4.00	4.00	9.00	21.000000	3.22750

4.00	4.00	9.00	42.000000	3.73743

4.00	4.00	9.00	63.000000	3.90342

4.00	4.00	9.00	84.000000	3.96312

4.00	4.00	9.00	105.000000	3.98560

4.00	4.00	9.00	126.000000	3.99429

4.00	4.00	9.00	147.000000	3.99771

4.00	4.00	9.00	168.000000	3.99908

4.00	4.00	9.00	189.000000	3.99962

4.00	4.00	9.00	210.000000	3.99985

4.00	4.00	9.00	231.000000	3.99994

4.00	4.00	9.00	252.000000	3.99997

4.00	4.00	10.00	21.000000	3.40682

4.00	4.00	10.00	42.000000	3.83646

4.00	4.00	10.00	63.000000	3.95094

4.00	4.00	10.00	84.000000	3.98469

4.00	4.00	10.00	105.000000	3.99511

4.00	4.00	10.00	126.000000	3.99841

4.00	4.00	10.00	147.000000	3.99948

4.00	4.00	10.00	168.000000	3.99983

4.00	4.00	10.00	189.000000	3.99994

4.00	4.00	10.00	210.000000	3.99998

4.00	4.00	10.00	231.000000	3.99999

4.00	4.00	10.00	252.000000	4.00000

4.00	5.00	1.00	2


4.00	7.00	7.00	21.000000	2.38621

4.00	7.00	7.00	42.000000	3.20615

4.00	7.00	7.00	63.000000	3.58476

4.00	7.00	7.00	84.000000	3.77605

4.00	7.00	7.00	105.000000	3.87694

4.00	7.00	7.00	126.000000	3.93151

4.00	7.00	7.00	147.000000	3.96152

4.00	7.00	7.00	168.000000	3.97823

4.00	7.00	7.00	189.000000	3.98761

4.00	7.00	7.00	210.000000	3.99292

4.00	7.00	7.00	231.000000	3.99594

4.00	7.00	7.00	252.000000	3.99766

4.00	7.00	8.00	21.000000	2.71317

4.00	7.00	8.00	42.000000	3.46734

4.00	7.00	8.00	63.000000	3.76381

4.00	7.00	8.00	84.000000	3.89168

4.00	7.00	8.00	105.000000	3.94930

4.00	7.00	8.00	126.000000	3.97594

4.00	7.00	8.00	147.000000	3.98847

4.00	7.00	8.00	168.000000	3.99443

4.00	7.00	8.00	189.000000	3.99729

4.00	7.00	8.00	210.000000	3.99868

4.00	7.00	8.00	231.000000	3.99935

4.00	7.00	8.00	252.000000	3.99968

4.00	7.00	9.00	21.000000	2.99203

4.00	7.00	9.00	42.000000	3.65537

4.00	7.00	9.00	63.000000	3.87294

4.00	7.00	9.00	84.000000	3.95142

4.00	7.00	9.00	105.000000	3.981


5.00	1.00	4.00	42.000000	4.26344

5.00	1.00	4.00	63.000000	4.41011

5.00	1.00	4.00	84.000000	4.52694

5.00	1.00	4.00	105.000000	4.61928

5.00	1.00	4.00	126.000000	4.69247

5.00	1.00	4.00	147.000000	4.75077

5.00	1.00	4.00	168.000000	4.79743

5.00	1.00	4.00	189.000000	4.83493

5.00	1.00	4.00	210.000000	4.86520

5.00	1.00	4.00	231.000000	4.88970

5.00	1.00	4.00	252.000000	4.90959

5.00	1.00	5.00	21.000000	4.18683

5.00	1.00	5.00	42.000000	4.42247

5.00	1.00	5.00	63.000000	4.58894

5.00	1.00	5.00	84.000000	4.70485

5.00	1.00	5.00	105.000000	4.78642

5.00	1.00	5.00	126.000000	4.84447

5.00	1.00	5.00	147.000000	4.88615

5.00	1.00	5.00	168.000000	4.91631

5.00	1.00	5.00	189.000000	4.93826

5.00	1.00	5.00	210.000000	4.95431

5.00	1.00	5.00	231.000000	4.96611

5.00	1.00	5.00	252.000000	4.97480

5.00	1.00	6.00	21.000000	4.29764

5.00	1.00	6.00	42.000000	4.56995

5.00	1.00	6.00	63.000000	4.73227

5.00	1.00	6.00	84.000000	4.83078

5.00	1.00	6.00	105.000000	4.89183

5.00	1.00	6.00	126.000000	4.93


5.00	4.00	9.00	105.000000	4.98392

5.00	4.00	9.00	126.000000	4.99363

5.00	4.00	9.00	147.000000	4.99745

5.00	4.00	9.00	168.000000	4.99897

5.00	4.00	9.00	189.000000	4.99958

5.00	4.00	9.00	210.000000	4.99983

5.00	4.00	9.00	231.000000	4.99993

5.00	4.00	9.00	252.000000	4.99997

5.00	4.00	10.00	21.000000	4.33818

5.00	4.00	10.00	42.000000	4.81740

5.00	4.00	10.00	63.000000	4.94520

5.00	4.00	10.00	84.000000	4.98289

5.00	4.00	10.00	105.000000	4.99453

5.00	4.00	10.00	126.000000	4.99823

5.00	4.00	10.00	147.000000	4.99942

5.00	4.00	10.00	168.000000	4.99981

5.00	4.00	10.00	189.000000	4.99994

5.00	4.00	10.00	210.000000	4.99998

5.00	4.00	10.00	231.000000	4.99999

5.00	4.00	10.00	252.000000	5.00000

5.00	5.00	1.00	21.000000	0.59243

5.00	5.00	1.00	42.000000	0.84401

5.00	5.00	1.00	63.000000	1.03780

5.00	5.00	1.00	84.000000	1.20100

5.00	5.00	1.00	105.000000	1.34427

5.00	5.00	1.00	126.000000	1.47310

5.00	5.00	1.00	147.000000	1.59079

5.00	5.00	1.00	168.000000	1.69951

5.00	5.00	1.00	


5.00	7.00	3.00	126.000000	3.34161

5.00	7.00	3.00	147.000000	3.55840

5.00	7.00	3.00	168.000000	3.74181

5.00	7.00	3.00	189.000000	3.89833

5.00	7.00	3.00	210.000000	4.03278

5.00	7.00	3.00	231.000000	4.14890

5.00	7.00	3.00	252.000000	4.24961

5.00	7.00	4.00	21.000000	1.72651

5.00	7.00	4.00	42.000000	2.59113

5.00	7.00	4.00	63.000000	3.15785

5.00	7.00	4.00	84.000000	3.56292

5.00	7.00	4.00	105.000000	3.86478

5.00	7.00	4.00	126.000000	4.09535

5.00	7.00	4.00	147.000000	4.27438

5.00	7.00	4.00	168.000000	4.41502

5.00	7.00	4.00	189.000000	4.52648

5.00	7.00	4.00	210.000000	4.61542

5.00	7.00	4.00	231.000000	4.68677

5.00	7.00	4.00	252.000000	4.74428

5.00	7.00	5.00	21.000000	2.25818

5.00	7.00	5.00	42.000000	3.20802

5.00	7.00	5.00	63.000000	3.77118

5.00	7.00	5.00	84.000000	4.13665

5.00	7.00	5.00	105.000000	4.38420

5.00	7.00	5.00	126.000000	4.55617

5.00	7.00	5.00	147.000000	4.67765

5.00	7.00	5.00	168.000000	4.76448

5.00	7.00	5.00	189.000000	4.82711

5.00	7.00	5.00	210.000000	4

5.00	10.00	1.00	126.000000	0.46314

5.00	10.00	1.00	147.000000	0.57129

5.00	10.00	1.00	168.000000	0.67733

5.00	10.00	1.00	189.000000	0.78074

5.00	10.00	1.00	210.000000	0.88125

5.00	10.00	1.00	231.000000	0.97878

5.00	10.00	1.00	252.000000	1.07335

5.00	10.00	2.00	21.000000	0.22819

5.00	10.00	2.00	42.000000	0.62272

5.00	10.00	2.00	63.000000	0.98240

5.00	10.00	2.00	84.000000	1.29966

5.00	10.00	2.00	105.000000	1.58064

5.00	10.00	2.00	126.000000	1.83143

5.00	10.00	2.00	147.000000	2.05694

5.00	10.00	2.00	168.000000	2.26097

5.00	10.00	2.00	189.000000	2.44652

5.00	10.00	2.00	210.000000	2.61602

5.00	10.00	2.00	231.000000	2.77141

5.00	10.00	2.00	252.000000	2.91431

5.00	10.00	3.00	21.000000	0.70531

5.00	10.00	3.00	42.000000	1.41917

5.00	10.00	3.00	63.000000	1.96730

5.00	10.00	3.00	84.000000	2.40373

5.00	10.00	3.00	105.000000	2.76039

5.00	10.00	3.00	126.000000	3.05710

5.00	10.00	3.00	147.000000	3.30707

5.00	10.00	3.00	168.000000	3.51966

5.00	10.00	3.00	189.000000	3.70180




6.00	4.00	1.00	168.000000	2.90475

6.00	4.00	1.00	189.000000	2.99492

6.00	4.00	1.00	210.000000	3.08026

6.00	4.00	1.00	231.000000	3.16129

6.00	4.00	1.00	252.000000	3.23842

6.00	4.00	2.00	21.000000	2.41521

6.00	4.00	2.00	42.000000	2.79821

6.00	4.00	2.00	63.000000	3.10112

6.00	4.00	2.00	84.000000	3.35303

6.00	4.00	2.00	105.000000	3.56895

6.00	4.00	2.00	126.000000	3.75769

6.00	4.00	2.00	147.000000	3.92492

6.00	4.00	2.00	168.000000	4.07458

6.00	4.00	2.00	189.000000	4.20955

6.00	4.00	2.00	210.000000	4.33201

6.00	4.00	2.00	231.000000	4.44366

6.00	4.00	2.00	252.000000	4.54587

6.00	4.00	3.00	21.000000	2.85856

6.00	4.00	3.00	42.000000	3.43259

6.00	4.00	3.00	63.000000	3.84583

6.00	4.00	3.00	84.000000	4.16629

6.00	4.00	3.00	105.000000	4.42438

6.00	4.00	3.00	126.000000	4.63708

6.00	4.00	3.00	147.000000	4.81512

6.00	4.00	3.00	168.000000	4.96580

6.00	4.00	3.00	189.000000	5.09441

6.00	4.00	3.00	210.000000	5.20491

6.00	4.00	3.00	231.000000	5.30035

6.00	4.00	3.00	252.000000	5


6.00	8.00	6.00	126.000000	5.77124

6.00	8.00	6.00	147.000000	5.85246

6.00	8.00	6.00	168.000000	5.90421

6.00	8.00	6.00	189.000000	5.93748

6.00	8.00	6.00	210.000000	5.95902

6.00	8.00	6.00	231.000000	5.97304

6.00	8.00	6.00	252.000000	5.98221

6.00	8.00	7.00	21.000000	3.85602

6.00	8.00	7.00	42.000000	4.95153

6.00	8.00	7.00	63.000000	5.45291

6.00	8.00	7.00	84.000000	5.70535

6.00	8.00	7.00	105.000000	5.83823

6.00	8.00	7.00	126.000000	5.91002

6.00	8.00	7.00	147.000000	5.94948

6.00	8.00	7.00	168.000000	5.97143

6.00	8.00	7.00	189.000000	5.98375

6.00	8.00	7.00	210.000000	5.99071

6.00	8.00	7.00	231.000000	5.99467

6.00	8.00	7.00	252.000000	5.99693

6.00	8.00	8.00	21.000000	4.29463

6.00	8.00	8.00	42.000000	5.29761

6.00	8.00	8.00	63.000000	5.68918

6.00	8.00	8.00	84.000000	5.85762

6.00	8.00	8.00	105.000000	5.93341

6.00	8.00	8.00	126.000000	5.96842

6.00	8.00	8.00	147.000000	5.98487

6.00	8.00	8.00	168.000000	5.99269

6.00	8.00	8.00	189.000000	5.99645

6.00	8.00	8.00	210.000000	5


7.00	1.00	9.00	105.000000	6.99093

7.00	1.00	9.00	126.000000	6.99638

7.00	1.00	9.00	147.000000	6.99854

7.00	1.00	9.00	168.000000	6.99941

7.00	1.00	9.00	189.000000	6.99976

7.00	1.00	9.00	210.000000	6.99990

7.00	1.00	9.00	231.000000	6.99996

7.00	1.00	9.00	252.000000	6.99998

7.00	1.00	10.00	21.000000	6.66120

7.00	1.00	10.00	42.000000	6.90082

7.00	1.00	10.00	63.000000	6.96952

7.00	1.00	10.00	84.000000	6.99036

7.00	1.00	10.00	105.000000	6.99689

7.00	1.00	10.00	126.000000	6.99899

7.00	1.00	10.00	147.000000	6.99967

7.00	1.00	10.00	168.000000	6.99989

7.00	1.00	10.00	189.000000	6.99996

7.00	1.00	10.00	210.000000	6.99999

7.00	1.00	10.00	231.000000	7.00000

7.00	1.00	10.00	252.000000	7.00000

7.00	2.00	1.00	21.000000	5.01660

7.00	2.00	1.00	42.000000	5.03344

7.00	2.00	1.00	63.000000	5.05241

7.00	2.00	1.00	84.000000	5.07485

7.00	2.00	1.00	105.000000	5.10063

7.00	2.00	1.00	126.000000	5.12908

7.00	2.00	1.00	147.000000	5.15947

7.00	2.00	1.00	168.000000	5.19117

7.00	2.00	1.00	


7.00	3.00	9.00	210.000000	6.99983

7.00	3.00	9.00	231.000000	6.99993

7.00	3.00	9.00	252.000000	6.99997

7.00	3.00	10.00	21.000000	6.33940

7.00	3.00	10.00	42.000000	6.81577

7.00	3.00	10.00	63.000000	6.94448

7.00	3.00	10.00	84.000000	6.98263

7.00	3.00	10.00	105.000000	6.99444

7.00	3.00	10.00	126.000000	6.99819

7.00	3.00	10.00	147.000000	6.99941

7.00	3.00	10.00	168.000000	6.99980

7.00	3.00	10.00	189.000000	6.99993

7.00	3.00	10.00	210.000000	6.99998

7.00	3.00	10.00	231.000000	6.99999

7.00	3.00	10.00	252.000000	7.00000

7.00	4.00	1.00	21.000000	3.04717

7.00	4.00	1.00	42.000000	3.14156

7.00	4.00	1.00	63.000000	3.25162

7.00	4.00	1.00	84.000000	3.36260

7.00	4.00	1.00	105.000000	3.47019

7.00	4.00	1.00	126.000000	3.57321

7.00	4.00	1.00	147.000000	3.67144

7.00	4.00	1.00	168.000000	3.76506

7.00	4.00	1.00	189.000000	3.85433

7.00	4.00	1.00	210.000000	3.93957

7.00	4.00	1.00	231.000000	4.02107

7.00	4.00	1.00	252.000000	4.09911

7.00	4.00	2.00	21.000000	3.28785

7.00	4.00	2.00	4


7.00	7.00	4.00	84.000000	5.29133

7.00	7.00	4.00	105.000000	5.65156

7.00	7.00	4.00	126.000000	5.92620

7.00	7.00	4.00	147.000000	6.13917

7.00	7.00	4.00	168.000000	6.30632

7.00	7.00	4.00	189.000000	6.43869

7.00	7.00	4.00	210.000000	6.54426

7.00	7.00	4.00	231.000000	6.62891

7.00	7.00	4.00	252.000000	6.69710

7.00	7.00	5.00	21.000000	3.72031

7.00	7.00	5.00	42.000000	4.86586

7.00	7.00	5.00	63.000000	5.53932

7.00	7.00	5.00	84.000000	5.97487

7.00	7.00	5.00	105.000000	6.26934

7.00	7.00	5.00	126.000000	6.47366

7.00	7.00	5.00	147.000000	6.61788

7.00	7.00	5.00	168.000000	6.72090

7.00	7.00	5.00	189.000000	6.79517

7.00	7.00	5.00	210.000000	6.84909

7.00	7.00	5.00	231.000000	6.88845

7.00	7.00	5.00	252.000000	6.91732

7.00	7.00	6.00	21.000000	4.30593

7.00	7.00	6.00	42.000000	5.46814

7.00	7.00	6.00	63.000000	6.07634

7.00	7.00	6.00	84.000000	6.42680

7.00	7.00	6.00	105.000000	6.63798

7.00	7.00	6.00	126.000000	6.76861

7.00	7.00	6.00	147.000000	6.85080

7.00	7.00	6.00	168.000000	6.

7.00	10.00	9.00	231.000000	6.99987

7.00	10.00	9.00	252.000000	6.99995

7.00	10.00	10.00	21.000000	5.76509

7.00	10.00	10.00	42.000000	6.65883

7.00	10.00	10.00	63.000000	6.89757

7.00	10.00	10.00	84.000000	6.96801

7.00	10.00	10.00	105.000000	6.98978

7.00	10.00	10.00	126.000000	6.99668

7.00	10.00	10.00	147.000000	6.99891

7.00	10.00	10.00	168.000000	6.99964

7.00	10.00	10.00	189.000000	6.99988

7.00	10.00	10.00	210.000000	6.99996

7.00	10.00	10.00	231.000000	6.99999

7.00	10.00	10.00	252.000000	7.00000

8.00	1.00	1.00	21.000000	7.00830

8.00	1.00	1.00	42.000000	7.01653

8.00	1.00	1.00	63.000000	7.02469

8.00	1.00	1.00	84.000000	7.03283

8.00	1.00	1.00	105.000000	7.04104

8.00	1.00	1.00	126.000000	7.04945

8.00	1.00	1.00	147.000000	7.05817

8.00	1.00	1.00	168.000000	7.06728

8.00	1.00	1.00	189.000000	7.07680

8.00	1.00	1.00	210.000000	7.08674

8.00	1.00	1.00	231.000000	7.09707

8.00	1.00	1.00	252.000000	7.10777

8.00	1.00	2.00	21.000000	7.00836

8.00	1.00	2.00	42.000000	7.02000

8.00

8.00	5.00	9.00	168.000000	7.99854

8.00	5.00	9.00	189.000000	7.99941

8.00	5.00	9.00	210.000000	7.99976

8.00	5.00	9.00	231.000000	7.99990

8.00	5.00	9.00	252.000000	7.99996

8.00	5.00	10.00	21.000000	7.07037

8.00	5.00	10.00	42.000000	7.74280

8.00	5.00	10.00	63.000000	7.92274

8.00	5.00	10.00	84.000000	7.97587

8.00	5.00	10.00	105.000000	7.99229

8.00	5.00	10.00	126.000000	7.99750

8.00	5.00	10.00	147.000000	7.99918

8.00	5.00	10.00	168.000000	7.99973

8.00	5.00	10.00	189.000000	7.99991

8.00	5.00	10.00	210.000000	7.99997

8.00	5.00	10.00	231.000000	7.99999

8.00	5.00	10.00	252.000000	8.00000

8.00	6.00	1.00	21.000000	2.20680

8.00	6.00	1.00	42.000000	2.46554

8.00	6.00	1.00	63.000000	2.69241

8.00	6.00	1.00	84.000000	2.89357

8.00	6.00	1.00	105.000000	3.07506

8.00	6.00	1.00	126.000000	3.24104

8.00	6.00	1.00	147.000000	3.39440

8.00	6.00	1.00	168.000000	3.53722

8.00	6.00	1.00	189.000000	3.67106

8.00	6.00	1.00	210.000000	3.79710

8.00	6.00	1.00	231.000000	3.91629

8.00	6.00	1.00	2


8.00	7.00	9.00	210.000000	7.99971

8.00	7.00	9.00	231.000000	7.99988

8.00	7.00	9.00	252.000000	7.99995

8.00	7.00	10.00	21.000000	6.89112

8.00	7.00	10.00	42.000000	7.69420

8.00	7.00	10.00	63.000000	7.90825

8.00	7.00	10.00	84.000000	7.97136

8.00	7.00	10.00	105.000000	7.99085

8.00	7.00	10.00	126.000000	7.99703

8.00	7.00	10.00	147.000000	7.99903

8.00	7.00	10.00	168.000000	7.99968

8.00	7.00	10.00	189.000000	7.99989

8.00	7.00	10.00	210.000000	7.99996

8.00	7.00	10.00	231.000000	7.99999

8.00	7.00	10.00	252.000000	8.00000

8.00	8.00	1.00	21.000000	0.94789

8.00	8.00	1.00	42.000000	1.35042

8.00	8.00	1.00	63.000000	1.66047

8.00	8.00	1.00	84.000000	1.92160

8.00	8.00	1.00	105.000000	2.15083

8.00	8.00	1.00	126.000000	2.35695

8.00	8.00	1.00	147.000000	2.54526

8.00	8.00	1.00	168.000000	2.71921

8.00	8.00	1.00	189.000000	2.88124

8.00	8.00	1.00	210.000000	3.03314

8.00	8.00	1.00	231.000000	3.17626

8.00	8.00	1.00	252.000000	3.31168

8.00	8.00	2.00	21.000000	1.84320

8.00	8.00	2.00	4

8.00	10.00	6.00	42.000000	6.04675

8.00	10.00	6.00	63.000000	6.82145

8.00	10.00	6.00	84.000000	7.26834

8.00	10.00	6.00	105.000000	7.53779

8.00	10.00	6.00	126.000000	7.70451

8.00	10.00	6.00	147.000000	7.80944

8.00	10.00	6.00	168.000000	7.87629

8.00	10.00	6.00	189.000000	7.91926

8.00	10.00	6.00	210.000000	7.94708

8.00	10.00	6.00	231.000000	7.96519

8.00	10.00	6.00	252.000000	7.97703

8.00	10.00	7.00	21.000000	5.22642

8.00	10.00	7.00	42.000000	6.64479

8.00	10.00	7.00	63.000000	7.29310

8.00	10.00	7.00	84.000000	7.61936

8.00	10.00	7.00	105.000000	7.79105

8.00	10.00	7.00	126.000000	7.88379

8.00	10.00	7.00	147.000000	7.93475

8.00	10.00	7.00	168.000000	7.96310

8.00	10.00	7.00	189.000000	7.97901

8.00	10.00	7.00	210.000000	7.98801

8.00	10.00	7.00	231.000000	7.99312

8.00	10.00	7.00	252.000000	7.99604

8.00	10.00	8.00	21.000000	5.79461

8.00	10.00	8.00	42.000000	7.09233

8.00	10.00	8.00	63.000000	7.59846

8.00	10.00	8.00	84.000000	7.81609

8.00	10.00	8.00	105.000000	7.91399

8.0

9.00	3.00	2.00	252.000000	7.57144

9.00	3.00	3.00	21.000000	6.22526

9.00	3.00	3.00	42.000000	6.60396

9.00	3.00	3.00	63.000000	6.93577

9.00	3.00	3.00	84.000000	7.21452

9.00	3.00	3.00	105.000000	7.44901

9.00	3.00	3.00	126.000000	7.64765

9.00	3.00	3.00	147.000000	7.81707

9.00	3.00	3.00	168.000000	7.96245

9.00	3.00	3.00	189.000000	8.08784

9.00	3.00	3.00	210.000000	8.19646

9.00	3.00	3.00	231.000000	8.29091

9.00	3.00	3.00	252.000000	8.37328

9.00	3.00	4.00	21.000000	6.51019

9.00	3.00	4.00	42.000000	7.07961

9.00	3.00	4.00	63.000000	7.50012

9.00	3.00	4.00	84.000000	7.81539

9.00	3.00	4.00	105.000000	8.05647

9.00	3.00	4.00	126.000000	8.24363

9.00	3.00	4.00	147.000000	8.39057

9.00	3.00	4.00	168.000000	8.50694

9.00	3.00	4.00	189.000000	8.59973

9.00	3.00	4.00	210.000000	8.67413

9.00	3.00	4.00	231.000000	8.73406

9.00	3.00	4.00	252.000000	8.78251

9.00	3.00	5.00	21.000000	6.84412

9.00	3.00	5.00	42.000000	7.53574

9.00	3.00	5.00	63.000000	7.97904

9.00	3.00	5.00	84.000000	8.2757

9.00	7.00	9.00	210.000000	8.99970

9.00	7.00	9.00	231.000000	8.99988

9.00	7.00	9.00	252.000000	8.99995

9.00	7.00	10.00	21.000000	7.82601

9.00	7.00	10.00	42.000000	8.67601

9.00	7.00	10.00	63.000000	8.90277

9.00	7.00	10.00	84.000000	8.96965

9.00	7.00	10.00	105.000000	8.99030

9.00	7.00	10.00	126.000000	8.99685

9.00	7.00	10.00	147.000000	8.99897

9.00	7.00	10.00	168.000000	8.99966

9.00	7.00	10.00	189.000000	8.99989

9.00	7.00	10.00	210.000000	8.99996

9.00	7.00	10.00	231.000000	8.99999

9.00	7.00	10.00	252.000000	9.00000

9.00	8.00	1.00	21.000000	1.59540

9.00	8.00	1.00	42.000000	2.00241

9.00	8.00	1.00	63.000000	2.32291

9.00	8.00	1.00	84.000000	2.59512

9.00	8.00	1.00	105.000000	2.83509

9.00	8.00	1.00	126.000000	3.05144

9.00	8.00	1.00	147.000000	3.24940

9.00	8.00	1.00	168.000000	3.43248

9.00	8.00	1.00	189.000000	3.60315

9.00	8.00	1.00	210.000000	3.76323

9.00	8.00	1.00	231.000000	3.91413

9.00	8.00	1.00	252.000000	4.05694

9.00	8.00	2.00	21.000000	2.50098

9.00	8.00	2.00	42


10.00	2.00	7.00	21.000000	8.82233

10.00	2.00	7.00	42.000000	9.38793

10.00	2.00	7.00	63.000000	9.67232

10.00	2.00	7.00	84.000000	9.82094

10.00	2.00	7.00	105.000000	9.90076

10.00	2.00	7.00	126.000000	9.94443

10.00	2.00	7.00	147.000000	9.96864

10.00	2.00	7.00	168.000000	9.98220

10.00	2.00	7.00	189.000000	9.98984

10.00	2.00	7.00	210.000000	9.99418

10.00	2.00	7.00	231.000000	9.99666

10.00	2.00	7.00	252.000000	9.99807

10.00	2.00	8.00	21.000000	9.03787

10.00	2.00	8.00	42.000000	9.58241

10.00	2.00	8.00	63.000000	9.81116

10.00	2.00	8.00	84.000000	9.91244

10.00	2.00	8.00	105.000000	9.95873

10.00	2.00	8.00	126.000000	9.98032

10.00	2.00	8.00	147.000000	9.99053

10.00	2.00	8.00	168.000000	9.99541

10.00	2.00	8.00	189.000000	9.99777

10.00	2.00	8.00	210.000000	9.99891

10.00	2.00	8.00	231.000000	9.99946

10.00	2.00	8.00	252.000000	9.99974

10.00	2.00	9.00	21.000000	9.23212

10.00	2.00	9.00	42.000000	9.72641

10.00	2.00	9.00	63.000000	9.89742

10.00	2.00	9.00	84.000000	9.96042

10.


10.00	4.00	10.00	105.000000	9.99234

10.00	4.00	10.00	126.000000	9.99751

10.00	4.00	10.00	147.000000	9.99918

10.00	4.00	10.00	168.000000	9.99973

10.00	4.00	10.00	189.000000	9.99991

10.00	4.00	10.00	210.000000	9.99997

10.00	4.00	10.00	231.000000	9.99999

10.00	4.00	10.00	252.000000	10.00000

10.00	5.00	1.00	21.000000	5.04650

10.00	5.00	1.00	42.000000	5.12952

10.00	5.00	1.00	63.000000	5.23898

10.00	5.00	1.00	84.000000	5.35793

10.00	5.00	1.00	105.000000	5.47862

10.00	5.00	1.00	126.000000	5.59769

10.00	5.00	1.00	147.000000	5.71368

10.00	5.00	1.00	168.000000	5.82596

10.00	5.00	1.00	189.000000	5.93434

10.00	5.00	1.00	210.000000	6.03881

10.00	5.00	1.00	231.000000	6.13948

10.00	5.00	1.00	252.000000	6.23649

10.00	5.00	2.00	21.000000	5.25651

10.00	5.00	2.00	42.000000	5.66086

10.00	5.00	2.00	63.000000	6.02483

10.00	5.00	2.00	84.000000	6.34394

10.00	5.00	2.00	105.000000	6.62546

10.00	5.00	2.00	126.000000	6.87601

10.00	5.00	2.00	147.000000	7.10080

10.00	5.00	2.00	168.000000


10.00	7.00	2.00	147.000000	6.44187

10.00	7.00	2.00	168.000000	6.69937

10.00	7.00	2.00	189.000000	6.93140

10.00	7.00	2.00	210.000000	7.14179

10.00	7.00	2.00	231.000000	7.33352

10.00	7.00	2.00	252.000000	7.50895

10.00	7.00	3.00	21.000000	4.59682

10.00	7.00	3.00	42.000000	5.59419

10.00	7.00	3.00	63.000000	6.30705

10.00	7.00	3.00	84.000000	6.85826

10.00	7.00	3.00	105.000000	7.30151

10.00	7.00	3.00	126.000000	7.66646

10.00	7.00	3.00	147.000000	7.97172

10.00	7.00	3.00	168.000000	8.22996

10.00	7.00	3.00	189.000000	8.45030

10.00	7.00	3.00	210.000000	8.63955

10.00	7.00	3.00	231.000000	8.80297

10.00	7.00	3.00	252.000000	8.94470

10.00	7.00	4.00	21.000000	5.38633

10.00	7.00	4.00	42.000000	6.60420

10.00	7.00	4.00	63.000000	7.40310

10.00	7.00	4.00	84.000000	7.97422

10.00	7.00	4.00	105.000000	8.39981

10.00	7.00	4.00	126.000000	8.72488

10.00	7.00	4.00	147.000000	8.97727

10.00	7.00	4.00	168.000000	9.17553

10.00	7.00	4.00	189.000000	9.33265

10.00	7.00	4.00	210.000000	9.45802


10.00	10.00	3.00	63.000000	5.52403

10.00	10.00	3.00	84.000000	6.19945

10.00	10.00	3.00	105.000000	6.73981

10.00	10.00	3.00	126.000000	7.18330

10.00	10.00	3.00	147.000000	7.55345

10.00	10.00	3.00	168.000000	7.86610

10.00	10.00	3.00	189.000000	8.13254

10.00	10.00	3.00	210.000000	8.36119

10.00	10.00	3.00	231.000000	8.55848

10.00	10.00	3.00	252.000000	8.72949

10.00	10.00	4.00	21.000000	4.38646

10.00	10.00	4.00	42.000000	5.89231

10.00	10.00	4.00	63.000000	6.86644

10.00	10.00	4.00	84.000000	7.55904

10.00	10.00	4.00	105.000000	8.07365

10.00	10.00	4.00	126.000000	8.46600

10.00	10.00	4.00	147.000000	8.77024

10.00	10.00	4.00	168.000000	9.00903

10.00	10.00	4.00	189.000000	9.19813

10.00	10.00	4.00	210.000000	9.34894

10.00	10.00	4.00	231.000000	9.46987

10.00	10.00	4.00	252.000000	9.56729

10.00	10.00	5.00	21.000000	5.31473

10.00	10.00	5.00	42.000000	6.95123

10.00	10.00	5.00	63.000000	7.91332

10.00	10.00	5.00	84.000000	8.53553

10.00	10.00	5.00	105.000000	8.95620

10.00	10.00

In [209]:
#for i in optionValues:
#    print(i,'\n')
#

In [208]:
#optionValues=bs(optionS)

In [43]:
myFunction(optionS)

array([ 3.,  5.,  7.,  9., 11., 13., 15., 17., 19.])

In [210]:
optionValues

array([2.94619295e-01, 5.32460451e-01, 7.18329824e-01, 8.46599840e-01,
       9.24809067e-01, 9.66943765e-01, 9.87001282e-01, 9.95437908e-01,
       9.98573436e-01, 9.99603104e-01, 9.26282218e-02, 3.66286816e-01,
       6.11419679e-01, 7.86589040e-01, 8.94897251e-01, 9.53657039e-01,
       9.81740667e-01, 9.93582795e-01, 9.97991337e-01, 9.99440736e-01,
       3.47894273e-02, 2.75897300e-01, 5.43267638e-01, 7.45665418e-01,
       8.73754438e-01, 9.44060503e-01, 9.77887264e-01, 9.92210680e-01,
       9.97557781e-01, 9.99319174e-01, 1.49171263e-02, 2.18565441e-01,
       4.93705469e-01, 7.14206946e-01, 8.57027760e-01, 9.36338037e-01,
       9.74752143e-01, 9.91085912e-01, 9.97200468e-01, 9.99218590e-01,
       7.05705116e-03, 1.78962749e-01, 4.55133406e-01, 6.88538094e-01,
       8.43045256e-01, 9.29790128e-01, 9.72069528e-01, 9.90117493e-01,
       9.96891459e-01, 9.99131320e-01, 3.60065948e-03, 1.50043397e-01,
       4.23818097e-01, 6.66816307e-01, 8.30961463e-01, 9.24061455e-01,
      

In [211]:
optionS.shape

(10,)

In [212]:
optionValues.shape

(1000,)

In [ ]:
optionArray=np.array((optionS,optionValues))

In [ ]:
optionArray.shape

In [ ]:
optionArray=optionArray.T

In [ ]:
type(optionArray)

In [ ]:
optionDF=pd.DataFrame(data=optionArray,columns=["S","Value"])

In [ ]:
optionDF['S']

In [ ]:
optionDF['Value']

In [ ]:
optionDF.shape

In [ ]:
underlying=(42/INCREMENT_S)

In [ ]:
lower=int(underlying-(10/INCREMENT_S))

In [ ]:
underlying

In [ ]:
upper=int(underlying+(10/INCREMENT_S))

In [ ]:
5/INCREMENT_S

In [ ]:
upper

In [ ]:
lower

In [ ]:
optionDF[optionDF['S']==1.0]

In [ ]:
optionDF

In [ ]:
plt.rcParams['figure.figsize'] = [7, 7] # This controls the picture size
fig, ax = plt.subplots()
ax.plot(optionDF['S'][lower:upper],optionDF['Value'][lower:upper])
#ax.plot()
ax.set(xlabel='underlying price', ylabel='option value',
       title='Call option value as function of current price')
ax.grid()

fig.savefig("test.png")
plt.show()

In [ ]:
x=(optionDF['S'][lower:upper])

In [ ]:
y=(optionDF['Value'][lower:upper])

In [ ]:
plt.plot(x,y)

In [ ]:
#optionDF.loc[['S']==42.0 ] #    [(S=37):(S=47)]

optionDF.loc[round(optionDF['S'],5)==42]

In [ ]:
optionDF.loc[optionDF['S']==1]

In [ ]:
optionDF.index

In [ ]:
optionDF.columns

In [ ]:
optionDF.iloc[41000]['Value']

In [ ]:
optionDF.iloc[41000]['S']

In [ ]:
optionDF.loc[optionDF.S==1]

In [ ]:
optionDF['Value']

In [ ]:
optionDF.loc[optionDF[S>41.000]]['Value']

In [ ]:
optionDF.dtypes

In [ ]:
optionDF